File: flaskApp-LayUI/flaskApp/views/file_list.py
# -*- coding: utf-8 -*-
from flask import make_response
from flaskApp.my_modules import mysqldb
from flaskApp import app
import json
import time
import re
import os


#http://localhost:3000/python/http_test?action=findData&whereStr=id=1 and name="xx"&fieldStr=field1,field2&prePageNum=10&currPage=1&sortStr=id ASC|DESC  //查询数据
#http://localhost:3000/python/http_test?action=insertData&dataArr=[{"class_name":"图库"},{"class_name":"图库"}]&fileArr=[file,file]  //上传文件
#http://localhost:3000/python/http_test?action=delData&whereJson={"url":[url1,url2]}  //删除文件


def operation(req):
    table_name = req.path[8:]
    dict_login = json.loads(req.cookies['logining'])

    # 查询数据
    def find_data():

        if 'whereStr' in req.args:
            str_where = req.args['whereStr']
        else:
            str_where = ''

        if 'fieldStr' in req.args:
            str_field = req.args['fieldStr']
        else:
            str_field = ''

        if 'sortStr' in req.args:
            str_sort = req.args['sortStr']
        else:
            str_sort = ''

        if 'prePageNum' in req.args:
            pre_page_num = int(req.args['prePageNum'])
        else:
            pre_page_num = 0

        if 'currPage' in req.args:
            curr_page = int(req.args['currPage'])
        else:
            curr_page = 0

        args = {'pre_page_num': pre_page_num, 'curr_page': curr_page, 'sort': str_sort}
        result = mysqldb.find_data(table_name, str_where, str_field, args)
        # print(result)

        if result:
            # 获取表头数据
            list_head = mysqldb.get_head('name="' + table_name + '"')

            if list_head and len(list_head) > 0:
                dict_json = {'code': 0, 'msg': '', 'count': result['count'], 'prePageNum': pre_page_num,
                             'currPage': curr_page, 'name': list_head[0]['name'], 'name_ch': list_head[0]['name_ch'],
                             'field_ch': list_head[0]['field_ch'], 'field_en': list_head[0]['field_en'],
                             'data_type': list_head[0]['data_type'], 'field_width': list_head[0]['field_width'],
                             'field_sort': list_head[0]['field_sort'], 'rows': result['rows']}
            else:
                dict_json = {'code': 0, 'msg': '', 'count': result['count'], 'prePageNum': pre_page_num,
                             'currPage': curr_page, 'name': '', 'name_ch': '', 'field_ch': '', 'field_en': '',
                             'data_type': '', 'field_width': '', 'field_sort': '', 'rows': result['rows']}

            return make_response(json.dumps(dict_json, ensure_ascii=False))
        else:
            return make_response('操作失败')


    # 上传文件
    def insert_data():
        if 'dataArr' in req.form:
            try:
                list_data = json.loads(req.form['dataArr'])
                if len(list_data) == 0:
                    return make_response('dataArr错误')
            except:
                return make_response('dataArr错误')
        else:
            return make_response('dataArr错误')

        upload_files = req.files.getlist("file")
        for i, file in enumerate(upload_files):
            now_time = time.strftime("%Y-%m-%d-%H-%M-%S", time.localtime()) + "-" + str(time.time())[11:15]
            now_path = os.path.dirname(os.path.dirname(__file__)) + "/static/uploadFile/" + now_time + "___" + file.filename
            if not os.path.exists(now_path):
                file.save(now_path)
            else:
                pass
            url = "/static/uploadFile/" + now_time + "___" + file.filename
            list_data[i]['name'] = file.filename
            list_data[i]['size'] = os.path.getsize(now_path)
            list_data[i]['url'] = url
            list_data[i]['create_name'] = dict_login['username']
            list_data[i]['create_time'] = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())

        # print(list_data)
        result = mysqldb.insert_data(table_name, list_data)
        # print(result)

        if result:
            # 操作记录
            content = 'dataArr=' + re.sub(r'\"', "'", json.dumps(list_data, ensure_ascii=False))
            dict_record = {'username': dict_login['username'], 'dbName': table_name, 'action': '上传文件', 'content': content, 'os': dict_login['os'], 'px': dict_login['px'], 'ip': req.remote_addr, 'time': time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())}
            mysqldb.set_record(dict_record)

            return make_response('操作成功')
        else:
            return make_response('操作失败')


    # 删除文件
    def del_data():
        if 'whereJson' in req.form:
            try:
                dict_where = json.loads(req.form['whereJson'])
                if not dict_where['url']:
                    return make_response('whereJson错误')
            except:
                return make_response('whereJson错误')
        else:
            return make_response('whereJson错误')

        result = mysqldb.del_data(table_name, dict_where)
        # print(result)

        if result:
            for url in dict_where['url']:
                file_path = os.path.normpath(os.path.dirname(os.path.dirname(__file__)) + url)
                if os.path.exists(file_path):
                    os.remove(file_path)
                else:
                    pass

            # 操作记录
            content = 'whereJson=' + re.sub(r'\"', "'", json.dumps(dict_where, ensure_ascii=False))
            dict_record = {'username': dict_login['username'], 'dbName': table_name, 'action': '删除文件', 'content': content, 'os': dict_login['os'], 'px': dict_login['px'], 'ip': req.remote_addr, 'time': time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())}
            mysqldb.set_record(dict_record)

            return make_response('操作成功')
        else:
            return make_response('操作失败')

    # GET请求
    if req.method == 'GET':
        print(req.args)

        # 判断权限
        if not mysqldb.get_power(dict_login['username'], dict_login['hash'], table_name, req.args['action']):
            return make_response('没有权限')

        if req.args['action'] == 'findData':
            return find_data()
        else:
            return make_response('action错误')

    # POST请求
    if req.method == 'POST':
        print(req.form)

        # 判断权限
        if not mysqldb.get_power(dict_login['username'], dict_login['hash'], table_name, req.form['action']):
            return make_response('没有权限')

        if req.form['action'] == 'insertData':
            return insert_data()
        elif req.form['action'] == 'delData':
            return del_data()
        else:
            return make_response('action错误')

    return make_response('action错误')

File: nodeApp-LayUI/routes/fileList.js
const express = require('express');
const router = express.Router();
const db = require("../my_modules/db.js");
const fs = require("fs");
const folder = require("../my_modules/folder");
const path = require('path');
const moment = require('moment');
const multer  = require('multer');
const upload = multer({ dest: '../public/uploadFile/',limits: { fileSize: 100*1024*1024, files:10 }});



// http://localhost:3000/nodejs/fileList?action=findData&searchJson={"class_name":"图库"}&fieldJson={"name":1}&prePageNum=10&currPage=1&sortJson={"key":1}  //查询文件
// http://localhost:3000/nodejs/fileList?action=insertData&dataJson={"class_name":"图库"}&fileArr=[file,file]  //上传文件 
// http://localhost:3000/nodejs/fileList?action=delData&dataJson={"url":{"$in":[url,url]}}  //删除文件


router.get('/', function(req, res, next) {
	
	var dbName = req.baseUrl.split("/")[2];
	var username = req.cookies["logining_node"].username;
	var hash = req.cookies["logining_node"].hash;
	
	//返回错误
    function returnErr (value) {
    	console.log(value);
		res.send(value);
    }
    
    //判断权限
	console.log(req.query);
	var action = req.query.action;
    db.getPower(username, hash, dbName, action, function(err, power){
	        if(err){
	            returnErr('getPowerError:'+ err);
			    return;
	        }
	        if(power == 0){
	        	res.send("没有权限");
	        	return;
	        }
	        
	        //分配方法
			switch (action){
				case "findData":
				      findData();
					  break;
		        default:
		              returnErr("action错误");
			}
    });
    
	
    
	//查询数据 
	function findData () {
		if(req.query.searchJson){
			try{
			  var whereJson = JSON.parse(req.query.searchJson);
			}catch(e){
				returnErr("searchJson错误");
				return;
			}
		}else{
			var whereJson = {};
		}
		
		if(req.query.sortJson){
			try{
			  var sortJson = JSON.parse(req.query.sortJson);
			}catch(e){
				returnErr("sortJson错误");
				return;
			}
		}else{
			var sortJson = {};
		}
		
		if(req.query.fieldJson){
			try{
			  var fieldJson = JSON.parse(req.query.fieldJson);
			}catch(e){
				returnErr("fieldJson错误");
				return;
			}
		}else{
			var fieldJson = {};
		}
		
		var prePageNum = req.query.prePageNum ? Number(req.query.prePageNum) : 0;
		var currPage = req.query.currPage ? Number(req.query.currPage) : 0;
		var args = {"prePageNum":prePageNum, "currPage":currPage, "sort":sortJson};
		db.findData(dbName, whereJson, fieldJson, args, function(err, result, count){
		        if(err){
		            returnErr('findDataError:'+ err);
				    return;
		        }
		        //console.log(result);
		        console.log("查询了" + result.length +"条数据，" + "共" + count +"条数据");
		        //获取表头数据
		        db.getHead({"name": dbName},function(err2, headArr){
				        if(err2){
				            returnErr(db.getHead+ err2);
						    return;
				        }
				       //console.log(headArr[0]);
				       if(headArr.length > 0){
				         var jsonObj = {"code": 0, "msg" : "", "count" : count, "prePageNum": prePageNum, "currPage": currPage, "name_ch": headArr[0].name_ch, "name": headArr[0].name, "field_ch": headArr[0].field_ch, "field_en": headArr[0].field_en, "data_type": headArr[0].data_type, "field_width": headArr[0].field_width, "rows": result};
				       }else{
				         var jsonObj = {"code": 0, "msg" : "", "count" : count, "prePageNum": prePageNum, "currPage": currPage, "name_ch": "", "name": "", "field_ch": "", "field_en": "", "data_type": "", "field_width": "", "rows": result};
				       }
				       //console.log(jsonObj);
				       res.send(jsonObj);
				});
		});
	}
	
});
	



router.post('/', upload.array('file'), function(req, res, next) {
	
	var dbName = req.baseUrl.split("/")[2];
	var username = req.cookies["logining_node"].username;
	var hash = req.cookies["logining_node"].hash;
	
	//返回错误
    function returnErr (value) {
    	console.log(value);
		res.send(value);
    }
    
    //判断权限
	console.log(req.body);
	var action = req.body.action;
	db.getPower(username, hash, dbName, action, function(err, power){
	        if(err){
	            returnErr('getPowerError:'+ err);
			    return;
	        }
	        if(power == 0){
	        	res.send("没有权限");
	        	return;
	        }
	        //分配方法
			switch (action){
			    case "insertData":
				      insertData();
					  break;
				case "delData":
				      delData();
					  break;
		        default:
		              returnErr("action错误");
			}
    });
    
	

	
	
	//上传文件
	function insertData () {
		if(req.body.dataJson){
			try{
			    var dataJson = JSON.parse(req.body.dataJson);
			    dataJson = {"class_name": dataJson.class_name};
			}catch(e){
				returnErr("dataJson错误");
				return;
			}
		}else{
			returnErr("dataJson错误");
			return;
		}
		if(req.files.length == 0){
			returnErr("上传文件不能为空");
			return;
		}
		if(!dataJson.class_name){
			returnErr("分类不能为空");
			return;
		}
		
		console.log(req.files);
		var newpathArr = [];
		var dataArr = [];
		for(let i=0; i<req.files.length; i++){
		   		var oldpath = req.files[i].path;
				var _url = "/public/uploadFile/" + moment().format("YYYY-MM-DD-HH-mm-ss-SS") + "___" + req.files[i].originalname;
			    var newpath = path.join(path.dirname(__dirname), _url);
			    console.log(newpath);
                try{
                  fs.renameSync(oldpath, newpath);
                }catch(err){
					returnErr('renameSync:'+ err);
					return;
				}
                newpathArr.push(newpath);
                dataJson.name = req.files[i].originalname;
                dataJson.size = req.files[i].size;
                dataJson.url = _url;
                dataJson.create_name = req.cookies["logining_node"].username;
                dataJson.create_time = moment().format("YYYY-MM-DD HH:mm:ss");
				var obj = JSON.parse(JSON.stringify(dataJson));
			    dataArr.push(obj);

		}
		console.log(dataArr);
        db.insertData(dbName,dataArr,function(err,result){
	        if(err){
	            for(let i=0; i<newpathArr.length; i++){
	               fs.unlink(newpathArr[i],function(err2){
	                 if(err2){console.log('unlinkError:'+ err2)}
	               });
	            }
		        returnErr('insertDataError:'+ err);
			    return;
	        }
	        console.log("上传了" + result.result.n +"个文件");
            if(result.result.n > 0){
	        	res.send("上传成功");
	        }else{
	        	res.send("上传失败");
	        }
            
            //删除缓存文件夹
//          folder.rmdirSync("../uploadFile",function(e){
//          	if(e){
//          	  console.log("folder.rmdirSync"+e);
//          	}
//			});
			
			//创建文件夹
//          folder.mkdirSync("../uploadFile2",function(e){
//          	if(e){
//          	  console.log("folder.rmdirSync"+e);
//          	}
//			});

            //操作记录
	        var content = "dataArr=" + JSON.stringify(dataArr).replace(/"/g,"'");
	        var recordJson = {"username":username, "dbName":dbName, "action": "上传文件", "content": content, "os": req.cookies["logining_node"].os, "px": req.cookies["logining_node"].px, "ip": req.ip, "time": moment().format("YYYY-MM-DD HH:mm:ss")};
	        db.setRecord(recordJson, function(recordErr,recordResult){
		        if(recordErr){
		            returnErr('setRecordError:'+ recordErr);
				    return;
		        }
		        console.log("操作记录成功"+content);
		    });

            
		});
		
	}
	
	
	
  

    //删除文件 
	function delData () {
		if(req.body.dataJson){
			try{
			  var whereJson = JSON.parse(req.body.dataJson);
			  var urlArr = whereJson["url"]["$in"];
			  if(Array.isArray(urlArr) && urlArr.length > 0){
					//continue
			  }else{
					returnErr("dataJson错误");
					return;
			  }
			}catch(e){
				returnErr("delJson错误");
				return;
			}
		}else{
			returnErr("delJson错误");
			return;
		}
		console.log(whereJson);
		db.deleteData(dbName, whereJson, function(err,result){
		        if(err){
		            returnErr('deleteDataError:'+ err);
				    return;
		        }
		        for(let i=0; i<urlArr.length; i++){
		           var urlpath = path.join(path.dirname(__dirname), urlArr[i]);
	               fs.unlink(urlpath,function(err2){
	                 if(err2){console.log('unlinkError:'+ err2)}
	               });
	            }
		        console.log("删除了" + result.result.n +"个文件");
		        if(result.result.n > 0){
		        	res.send("操作成功");
		        }else{
		        	res.send("操作失败");
		        }
		        
		        //操作记录
		        var content = "dataJson=" + JSON.stringify(whereJson).replace(/"/g,"'");
		        var recordJson = {"username":username, "dbName":dbName, "action": "删除文件", "content": content, "os": req.cookies["logining_node"].os, "px": req.cookies["logining_node"].px, "ip": req.ip, "time": moment().format("YYYY-MM-DD HH:mm:ss")};
		        db.setRecord(recordJson, function(recordErr,recordResult){
			        if(recordErr){
			            returnErr('setRecordError:'+ recordErr);
					    return;
			        }
			        console.log("操作记录成功"+content);
			    });
		});
	}
	
	
	
	
	
	
	
});

	




module.exports = router;


